chalice/local.py

```diff
@@ -109,98 +128,251 @@ def create_lambda_event(self, method, path, headers, body=None):
+        statements = policy.get('Statement', [])
+        # Look for a statement that allows us to invoke the route_entry
+        # it will be in the form local:local/METHOD/uri the URI is always
+        # prefixed with / so we do not need to explicitly include it in our
+        # pattern.
+        # The method can be either an explicit method such as POST, GET, etc...
+        # or it can be a *. We also need to match for any prefix since the
+        # user can return any policy at all from the custom authorizer.
+        method_pattern = r'(?:\*|%s)' % route_entry.method
+        route_entry_pattern = r'^.*?/%s%s$' % (method_pattern,
+                                               route_entry.uri_pattern)
+        route_entry_re = re.compile(route_entry_pattern)
+        # Look through all statements for one that allows us to invoke this
+        # particular view function.
+        for statement in statements:
+            if statement.get('Effect') == 'Allow' and \
+               statement.get('Action') == 'execute-api:Invoke':
+                for resource in statement.get('Resource'):
+                    if route_entry_re.match(resource):
+                        return True
+        return False
+
+    def _authorizer_and_route_for_event(self, lambda_event):
+        # type: (EventType) -> Tuple[Optional[Any], Optional[RouteEntry]]
+        # Authorizer had to be made into an Any type since mypy couldn't
+        # detect that app.ChaliceAuthorizer was callable.
+        resource_path = lambda_event.get(
+            'requestContext', {}).get('resourcePath')
+        http_method = lambda_event['requestContext']['httpMethod']
+        try:
+            route_entry = self._app_object.routes[resource_path][http_method]
+        except KeyError:
+            # If there a key error is raised when trying to get the route entry
+            # then this route does not support this method. A method error
+            # will be raised by the chalice handler method. We can ignore it
+            # here by returning no authorizer to avoid duplicating the logic.
+            return None, None
+        authorizer = route_entry.authorizer
+        return authorizer, route_entry
+
+    def _update_lambda_event(self, lambda_event, auth_result):
+        # type: (EventType, ResponseType) -> EventType
+        auth_context = auth_result['context']
+        auth_context.update({
+            'principalId': auth_result['principalId']
+        })
+        lambda_event['requestContext']['authorizer'] = auth_context
+        return lambda_event
 
-    def __init__(self, request, client_address, server, app_object):
-        # type: (bytes, Tuple[str, int], HTTPServer, Chalice) -> None
-        self.app_object = app_object
+    def _prepare_authorizer_event(self, lambda_event):
+        # type: (EventType) -> EventType
+        """Translate event for an authorizer input."""
+        authorizer_event = lambda_event.copy()
```

Comment:
How nested is this structure? if it's more than one level then maybe you should deepcopy.